### PR TITLE
Refactor Windowed Reading GeoTiffs from S3 and Hdfs

### PR DESCRIPTION
**Pull request by echeipesh**
Supersedes https://github.com/geotrellis/geotrellis/pull/1763

Main thrust was to refactor `S3GeoTiffRDD` and `HadoopGeoTiffRDD` to both use:

```scala
trait GeoTiffReader[-O, K, V] extends Serializable {
  def readFully(byteReader: ByteReader, options: O): (K, V)
  def readWindow(byteReader: StreamingByteReader, pixelWindow: GridBounds, options: O): (K, V)
}
```

This centralized the GeoTiff reading code that was spread among: `GeoTiffS3InputFormattable`, `GeoTiffInputFormattable`, `GeoTiffInputFormat`, `MultibandGeoTiffInputFormat`, `TemporalGeoTiffInputFormat`, `TemporalMultibandGeoTiffInputFormat`, `TemporalMultibandGeoTiffS3InputFormat`, `TemporalGeoTiffS3InputFormat`, `MultibandGeoTiffS3InputFormat`, `GeoTiffS3InputFormat`.

This allows to handle base configurations in one place and flush our some errors related to that.
 - Propagated `chunkSize` in `S3GeoTiffRDD.Options`
 - Added `tiffExtensions` to `S3GeoTiffRDD.Options`
 - Added `partitionBytes` to `S3GeotiffRDD.Options`
 
Currently both still use some form of `InputFormat` as base way to instantiate records for full reads.

### CODE REVIEW

File: s3/src/main/scala/geotrellis/spark/io/s3/S3Client.scala [hunk 1]
```diff
@@ -121,80 +121,11 @@ object S3Client {
...
-    }
-  }
-  
-  def getObjectMetadata(getObjectMetadataRequest: GetObjectMetadataRequest): ObjectMetadata =
-    s3client.getObjectMetadata(getObjectMetadataRequest)
-
-  def setRegion(region: com.amazonaws.regions.Region): Unit = {
-    s3client.setRegion(region)
-  }
+//  def apply(
```
Comment (lossyrob) [inline s3/src/main/scala/geotrellis/spark/io/s3/S3Client.scala]: rm

File: s3/src/main/scala/geotrellis/spark/io/s3/S3InputFormat.scala [hunk 2]
```diff
@@ -77,6 +87,14 @@ abstract class S3InputFormat[K, V] extends InputFormat[K,V] with LazyLogging {
...
         .listObjectsIterator(request)
         .filter(!_.getKey.endsWith("/"))
+        .filter { obj =>
+          if (extensions.isEmpty)
+            true
+          else {
+            val key = obj.getKey
+            extensions.map(key.endsWith).reduce(_ || _)
+          }
+        }
```
Comment (lossyrob) [inline s3/src/main/scala/geotrellis/spark/io/s3/S3InputFormat.scala]: is this a bugfix?

File: s3-test/src/test/scala/geotrellis/spark/io/s3/GeoTiffS3InputFormatSpec.scala [hunk 3]
```diff
@@ -1,26 +1,28 @@
...
+//   override def createRecordReader(split: InputSplit, context: TaskAttemptContext) =
+//     new MockGeoTiffS3RecordReader(context)
+// }
 
-class MockGeoTiffS3RecordReader(context: TaskAttemptContext) extends GeoTiffS3RecordReader(context) {
-  override def getS3Client(credentials: AWSCredentials): S3Client = new MockS3Client
-}
+// class MockGeoTiffS3RecordReader(context: TaskAttemptContext) extends GeoTiffS3RecordReader(context) {
+//   override def getS3Client(credentials: AWSCredentials): S3Client = new MockS3Client
+// }
```
Comment (lossyrob) [inline s3-test/src/test/scala/geotrellis/spark/io/s3/GeoTiffS3InputFormatSpec.scala]: Remove unused.

File: s3-test/src/test/scala/geotrellis/spark/io/s3/TemporalGeoTiffS3InputFormatSpec.scala [hunk 4]
```diff
@@ -21,15 +21,15 @@ import java.time.format.DateTimeFormatter
...
-class MockTemporalGeoTiffS3InputFormat extends TemporalGeoTiffS3InputFormat {
-  override def getS3Client(credentials: AWSCredentials): S3Client = new MockS3Client
-  override def createRecordReader(split: InputSplit, context: TaskAttemptContext) =
-    new MockTemporalGeoTiffS3RecordReader(context)
-}
+// class MockTemporalGeoTiffS3InputFormat extends TemporalGeoTiffS3InputFormat {
+//   override def getS3Client(credentials: AWSCredentials): S3Client = new MockS3Client
+//   override def createRecordReader(split: InputSplit, context: TaskAttemptContext) =
+//     new MockTemporalGeoTiffS3RecordReader(context)
+// }
```
Comment (lossyrob) [inline s3-test/src/test/scala/geotrellis/spark/io/s3/TemporalGeoTiffS3InputFormatSpec.scala]: Remove unused.

File: s3/src/main/scala/geotrellis/spark/io/s3/MultibandGeoTiffS3InputFormat.scala [hunk 5]
```diff
@@ -8,10 +8,10 @@ import org.apache.hadoop.mapreduce.{InputSplit, TaskAttemptContext}
 /** Read multi band GeoTiff from S3 */
 class MultibandGeoTiffS3InputFormat extends S3InputFormat[ProjectedExtent, MultibandTile] {
```
Comment (lossyrob) [inline s3/src/main/scala/geotrellis/spark/io/s3/MultibandGeoTiffS3InputFormat.scala]: Seems like these InputFormats are no longer used, and we're just reading in the bytes. Should we delete?

File: s3/src/main/scala/geotrellis/spark/io/s3/S3InputFormat.scala (hunk 2 above)
Comment (echeipesh) [inline s3/src/main/scala/geotrellis/spark/io/s3/S3InputFormat.scala]: No, S3InputFormat didn't have a way to filter by extension like hadoop input formats. This is addition to bring options of `S3GeoTiffRDD` and `HadoopGeoTiffRDD` to parity

**Pull request merged**